src/devices/Tlc1543/Tlc1543.cs

```diff
@@ -0,0 +1,289 @@
+
+        /// <summary>
+        /// Reads sensor value
+        /// </summary>
+        /// <param name="channelNumber">Channel to be read</param>
+        /// <returns>A 10 bit value corresponding to relative voltage level on specified device channel</returns>
+        public int ReadChannel(Channel channelNumber)
+        {
+            Read((int)channelNumber);
+            return Read((int)ChargeChannel);
+        }
+
+        /// <summary>
+        /// Reads sensor values into an List
+        /// </summary>
+        /// <param name="channelList">List of channels to read</param>
+        /// <returns>List of 10 bit values corresponding to relative voltage level on specified device channels</returns>
+        public List<int> ReadChannels(List<Channel> channelList)
+        {
+            List<int> values = new List<int>(channelList.Count);
+            Read((int)channelList[0]);
+            for (int i = 1; i < channelList.Count; i++)
+            {
+                values.Add(Read((int)channelList[i]));
+            }
+
+            values.Add(Read((int)ChargeChannel));
+            return values;
+        }
+
+        /// <summary>
+        /// Reads all sensor values into an List
+        /// </summary>
+        /// <returns>List of 10 bit values corresponding to relative voltage level on all channels</returns>
+        public List<int> ReadAll()
+        {
+            List<int> values = new List<int>(_channels);
+            Read(0);
+            for (int i = 1; i < values.Count; i++)
+            {
+                values.Add(Read(i));
+            }
+
+            values.Add(Read((int)ChargeChannel));
+            return values;
+        }
+
+        /// <summary>
+        /// Mode 1 (Fast mode):
+        /// <br>10 clock transfer with !CS high </br>
+        /// <br>Figure 9 in datasheet</br>
+        /// </summary>
+        /// <param name="channelNumber">Channel to read</param>
+        /// <returns>10 bit value corresponding to relative voltage level on channel</returns>
+        public int Read(int channelNumber)
+        {
+            int value = 0;
+            _digital.Write(_chipSelect, 0);
+            DelayHelper.DelayMicroseconds(1, false);
+            for (int i = 0; i < 10; i++)
+            {
+                if (i < 4)
+                {
+                    // send 4-bit Address
+                    if ((channelNumber >> (3 - i) & 0x01) != 0)
+                    {
+                        _digital.Write(_address, 1);
+                    }
+                    else
+                    {
+                        _digital.Write(_address, 0);
+                    }
+                }
+
+                // read 10-bit data
+                value <<= 1;
+                if (_digital.Read(_dataOut) == PinValue.High)
+                {
+                    value |= 0x01;
+                }
+
+                _digital.Write(_inputOutputClock, 1);
+                _digital.Write(_inputOutputClock, 0);

```

Comment:
Don't you have to wait a bit?

Reply:
I would have added delays there but simply rising edge (same with falling one) is 2 times as long as nominal time for the whole IOCLK period. Controller.Write is slow enough that adding any delay (even 1 tick long) doubles the length of IOCLK period and having that 10 to 20 times per conversions adds up quickly for performance hit that makes whole ADC polling one channel in ~1kHz speeds (instead of nominal 2MHz)

Reply:
What would happen on a real time processor like an ESP32 if we migrate this code into nanoFramework?
And that's then one more element to use SPI as well, this then will be done automatically by the clock pin

Reply:
interesting case... I think DelayHelper with no interrupts as parameter should work fine here though